codebasin/config.py

```diff
@@ -270,6 +306,225 @@ def load_codebase(config, rootdir):
     return codebase
 
 
+_importcfg = None
+def load_importcfg():
+    """
+    Load the import configuration file, if it exists.
+    """
+    global _importcfg
+    _importcfg = collections.defaultdict(list)
+    path = ".cbi/importcfg.json"
+    if os.path.exists(path):
+        log.info(f"Found import configuration file at {path}")
+        with open(path, "r") as f:
+            try:
+                _importcfg_json = json.load(f)
+                for compiler in _importcfg_json["compilers"]:
+                    _importcfg[compiler["name"]] = compiler["options"]
+            except BaseException:
+                log.error("importcfg file failed validation")
+
+
+class Compiler(object):
+    """
+    Represents the behavior of a specific compiler, including:
+    - The number of passes it performs.
+    - Implicitly defined macros, which may be flag-dependent.
+    """
+
+    def __init__(self, args):
+        self.name = os.path.basename(args[0])
+        self.args = args
+        self.passes = set(["default"])
+
+        # Check for any user-defined compiler behavior.
+        # Currently, users can only override default defines.
+        if _importcfg is None:
+            load_importcfg()
+        self.defines = extract_defines(_importcfg[self.name])
+
+    def get_passes(self):
+        return self.passes.copy()
+
+    def get_defines(self, pass_):
+        return self.defines.copy()
+
+    def get_include_paths(self, pass_):
+        return []
+
+    def get_include_files(self, pass_):
+        return []
+
+    def has_implicit_behavior(self, pass_):
+        return (self.get_defines(pass_) or
+                self.get_include_paths(pass_) or
+                self.get_include_files(pass_))
+
+    def get_configuration(self, pass_):
+        defines = self.get_defines(pass_)
+        include_paths = self.get_include_paths(pass_)
+        include_files = self.get_include_files(pass_)
+        return {"defines": defines,
+                "include_paths": include_paths,
+                "include_files": include_files}
+
+
+class ClangCompiler(Compiler):
+    """
+    Represents the behavior of Clang-based compilers.
+    """
+
+    device_passes = [
+        "spir64",
+        "x86_64",
+        "spir64_x86_64",
+        "spir64_gen",
+        "spir64_fpga"
+    ]
+
+    def __init__(self, args):
+        super().__init__(args)
+
+        self.sycl = False
+        self.omp = False
+        sycl_targets = []
+
+        for arg in args:
+
+            if arg == "-fsycl":
+                self.sycl = True
+                continue
+
+            m = re.search("-fsycl-targets=", arg)
+            if m:
+                sycl_targets = arg.split("=")[1].split(",")
+                self.passes |= set(sycl_targets)
+                continue
+
+            if arg in ["-fopenmp", "-fiopenmp", "-qopenmp"]:
```

Comment:
It does make things a little different, but would you rather work with `set` instead?

i.e.

```python
args = set(args)

openmp_args = ["-fopenmp", "-fiopenmp", "-qopenmp"]

if args.intersection(openmp_args):
   self.omp = True
```

This would also extend to other checks.

Reply:
I agree this looks nicer for the current checks, but I think we should probably keep things as a list.  Converting to a set representation would strip all the positional information, and I'm worried we might regret that later.

The compilation database format is a bit loose about what an argument actually is, and compilers are similarly pretty loose about how they interpret arguments anyway. The upshot is that sometimes we see argument lists like `["-D", "_OPENMP"]` and sometimes we see argument lists like `["-D _OPENMP"]` or `["-D_OPENMP"]`.

Because of this weirdness, I think there may be (future) cases where we need consider pairs of neighboring arguments.